astropy/cosmology/tests/test_connect.py

```diff
@@ -2,112 +2,56 @@
+
+###############################################################################
+# SETUP
+
 cosmo_instances = cosmology.parameters.available
 readwrite_formats = ["json"]
 tofrom_formats = [("mapping", dict)]  # (format, data type)
 
-
 ###############################################################################
-# Setup
-
-def read_json(filename, **kwargs):
-    with open(filename, "r") as file:
-        data = file.read()
-    mapping = json.loads(data)  # parse json mappable to dict
-    # deserialize Quantity
-    for k, v in mapping.items():
-        if isinstance(v, dict) and "value" in v and "unit" in v:
-            mapping[k] = u.Quantity(v["value"], v["unit"])
-    for k, v in mapping.get("meta", {}).items():  # also the metadata
-        if isinstance(v, dict) and "value" in v and "unit" in v:
-            mapping["meta"][k] = u.Quantity(v["value"], v["unit"])
-    return Cosmology.from_format(mapping, **kwargs)
-
-
-def write_json(cosmology, file, *, overwrite=False):
-    """Write Cosmology to JSON.
-
-    Parameters
-    ----------
-    cosmology : `astropy.cosmology.Cosmology` subclass instance
-    file : path-like or file-like
-    overwrite : bool (optional, keyword-only)
-    """
-    data = cosmology.to_format("mapping")  # start by turning into dict
-    data["cosmology"] = data["cosmology"].__qualname__  # class -> str
-    # serialize Quantity
-    for k, v in data.items():
-        if isinstance(v, u.Quantity):
-            data[k] = {"value": v.value.tolist(),
-                       "unit": str(v.unit)}
-    for k, v in data.get("meta", {}).items():  # also serialize the metadata
-        if isinstance(v, u.Quantity):
-            data["meta"][k] = {"value": v.value.tolist(),
-                               "unit": str(v.unit)}
-
-    # check that file exists and whether to overwrite.
-    if os.path.exists(file) and not overwrite:
-        raise IOError(f"{file} exists. Set 'overwrite' to write over.")
-    with open(file, "w") as write_file:
-        json.dump(data, write_file)
-
-
-def json_identify(origin, filepath, fileobj, *args, **kwargs):
-    return filepath is not None and filepath.endswith(".json")
-
-
-def setup_module(module):
-    """Setup module for tests."""
-    io_registry.register_reader("json", Cosmology, read_json)
-    io_registry.register_writer("json", Cosmology, write_json)
-    io_registry.register_identifier("json", Cosmology, json_identify)
-
-
-def teardown_module(module):
-    """clean up module after tests."""
-    io_registry.unregister_reader("json", Cosmology)
-    io_registry.unregister_writer("json", Cosmology)
-    io_registry.unregister_identifier("json", Cosmology)
 
 
-###############################################################################
-# Tests
+class ReadWriteTestMixin:
+    """Tests for a CosmologyRead/Write on a Cosmology."""
 
-class TestReadWriteCosmology:
+    @pytest.fixture(scope="class", autouse=True)
```

Comment:
Didn't know about this one. Out of curiosity, any reason not to use `setup_class` and `teardown_class`? Are you avoiding those to avoid forcing `super()` calls when actually used as a mixin?

Reply:
Yeah, it's mostly that.
And since these are "mixin" it helps keep the class orthogonal.